### DISCUSSION TITLE
FFI C++ examples under Windows 10 with Visual Studio 2022

### DISCUSSION THREAD

**Discussion by jlfaucher** [unicode-org/icu4x#2780] [Q&A] [answered]
Hello
I'm testing the FFI C++ examples under Windows 10 with Visual Studio 2022.
I run this command to build the library (from source v1.0.0)
`cargo build -p icu_capi_staticlib --features cpp_default`

and then this command from the example directory
`cl /std:c++20 /EHsc test.cpp ..\..\..\..\..\target\debug\icu_capi_staticlib.lib userenv.lib advapi32.lib ws2_32.lib bcrypt.lib`
(I use std:c++20 because the compiler says so)

Some examples work good, some doesn't.
I'm curious to know if the following problems are just me...

1)
When running locale/test, I get
`Created a locale: "s-ES"`
`Locale did not match expected: "es-ES"`
("s-ES" is truncated, works good under MacOs)

2)
When running the segmenter example, I get several warnings like that:
`2022-10-24 14:21:05,150 WARN [icu_provider::error] ICU4X data error: Missing data for locale (key: segmenter/dictionary@1, request: km)``
(same warnings under MacOs)

Thanks for your great work!
Jean-Louis

**Comment by sffc** [comment, marked as answer]
(1) looks odd. Why would the first character of the string be dropped? @Manishearth (or CC others)

(2) is a known issue. The warnings are harmless and can be ignored. I opened https://github.com/unicode-org/icu4x/issues/2781 to print better diagnostics and avoid printing noisy warnings.

**Reply by Manishearth** [reply]
> (1) looks odd. Why would the first character of the string be dropped? @Manishearth (or CC others)

Perhaps it's an ABI thing? We might be passing string parameters in a way that doesn't quite work on MSVC (i.e. our Rust-side extern C signatures and C-side headers are not quite compatible). But what we do here is pretty simple so I'd be rather surprised.

**Reply by Manishearth** [reply]
Either way I'd expect things to be mangled _way_ more in that case. So i'm not sure how that happened.

Worth adding a println to the locale constructor to check what string it _gets_ at least.

**Reply by jlfaucher** [reply]
About point (1)
Put a breakpoint in test_locale on the first line.
`  std::string_view actualString = locale.to_string().ok().value();`

Step into, step out several times.
All is good until a destructor is called.
See the attached file.
[debug-2780.txt](https://github.com/unicode-org/icu4x/files/9856613/debug-2780.txt)

**Reply by jlfaucher** [reply]
Updated attached file, the call stack was not complete.
[debug-2780-v2.txt](https://github.com/unicode-org/icu4x/files/9856634/debug-2780-v2.txt)

**Reply by sffc** [reply]
Hmm. It looks like the string gets a NUL byte added to it at some point.

@jefgen, this is a MSVC failure in ICU4X. Do you know someone who could take a look?

**Reply by jlfaucher** [reply]
For information, the C example works correctly.
So the problem seems limited to the management of temporary buffer by Diplomat ?

**Reply by jlfaucher** [reply]
It's a bad usage of `std::string_view` in the test case.
`::std:string_view` must not be used to refer a temporary string.
Use `std::string` instead.

current: `std::string_view actualString = locale.to_string().ok().value();`
to be: `std::string actualString = locale.to_string().ok().value();`

With this change, the test case works correctly when compiling with Visual Studio 2022.

**Reply by sffc** [reply]
Oh, yeah, that would do it! Good catch.

Want to open a PR to fix it?

**Reply by jlfaucher** [reply]
ok, I will.
